Event loop should attempt to run idle callbacks at the end of each task

rniwa opened the pull request:
#### 374b4feb0aeff3225da6e6f25579d7b460f1363e
<pre>
Event loop should attempt to run idle callbacks at the end of each task
<a href="https://bugs.webkit.org/show_bug.cgi?id=260115">https://bugs.webkit.org/show_bug.cgi?id=260115</a>

Reviewed by Wenson Hsieh.

This PR implements the step 8 in the HTML5 event loop processing model:
<a href="https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model">https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model</a>

Specifically, after executing each task in WindowEventLoop, we check if there are any tasks or microtasks
scheduled in the event loop. If there are none, we attempt to run idle callbacks.

This PR also fixes IdleDeadline::timeRemaining to subtract the current time before reducing the resolution
as it turns out that reducing the resolution before subtraction can generate the remaining time that is
slightly longer than the actual time remaining. e.g. 50.000000002ms as opposed to 50ms.

In addition, this PR updates WindowEventLoop::computeIdleDeadline() to take into account the next time
a timer associated with this event loop fires. This is computed &amp; cached by EventLoop::nextTimerFiringTime.

Finally, this PR also fixes a bug in Page::scheduleRenderingUpdateInternal() that it was sometimes calling
WindowEventLoop::didScheduleRenderingUpdate with a rendering update time in the past.

* LayoutTests/http/tests/eventloop/idle-callback-not-affected-by-timers-in-cross-origin-expected.txt: Added.
* LayoutTests/http/tests/eventloop/idle-callback-not-affected-by-timers-in-cross-origin.html: Added.
* LayoutTests/http/tests/eventloop/resources/idle-callback-timer-helper.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/requestidlecallback/deadline-max-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/requestidlecallback/deadline-max-rAF-dynamic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/requestidlecallback/deadline-max-timeout-dynamic-expected.txt:
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/requestidlecallback/requestidlecallback-in-page-cache.html: This test was relying on the idle
callback in the main frame to be invoked after idle callbacks in subframes. Since this is not guaranteed,
* Source/WebCore/dom/EventLoop.cpp:
(WebCore::EventLoop::scheduleTask):
(WebCore::EventLoop::removeScheduledTimer):
(WebCore::EventLoop::scheduleRepeatingTask):
(WebCore::EventLoop::removeRepeatingTimer):
(WebCore::EventLoop::findMatchingAssociatedContext): Added.
(WebCore::EventLoop::nextTimerFiringTime const): Added.
(WebCore::EventLoopTaskGroup::suspend):
(WebCore::EventLoopTaskGroup::resume):
(WebCore::EventLoopTaskGroup::setTimerAlignment):
(WebCore::EventLoopTaskGroup::didChangeTimerAlignmentInterval):
(WebCore::EventLoopTaskGroup::setTimerHasReachedMaxNestingLevel):
(WebCore::EventLoopTaskGroup::adjustTimerNextFireTime):
(WebCore::EventLoopTaskGroup::adjustTimerRepeatInterval):
* Source/WebCore/dom/EventLoop.h:
(WebCore::EventLoop::invalidateNextTimerFiringTimeCache): Added.
* Source/WebCore/dom/IdleCallbackController.h:
(WebCore::IdleCallbackController::isEmpty):
* Source/WebCore/dom/IdleDeadline.cpp:
(WebCore::IdleDeadline::timeRemaining const):
* Source/WebCore/dom/WindowEventLoop.cpp:
(WebCore::WindowEventLoop::computeIdleDeadline):
(WebCore::WindowEventLoop::didReachTimeToRun):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::scheduleRenderingUpdateInternal):
* Source/WebCore/platform/Timer.h:
(WebCore::TimerBase::nextFireTime const):

Canonical link: <a href="https://commits.webkit.org/266877@main">https://commits.webkit.org/266877@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/027ff49c718ed0f43b3c29a66660c9c3f5e5af08

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15018 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15323 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15677 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16774 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14130 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15157 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17843 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15423 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/16764 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15199 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15674 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12763 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17503 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12954 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13549 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/20519 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14032 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13717 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16960 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14276 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12080 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13556 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/13406 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17892 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1806 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14117 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/dom/EventLoop.cpp:
```diff
@@ -39,6 +39,7 @@ class EventLoopTimer final : public RefCounted<EventLoopTimer>, public TimerBase
 
     Type type() const { return m_type; }
     EventLoopTaskGroup* group() const { return m_task ? m_task->group() : nullptr; }
+    bool isSuspended() { return m_suspended; }
```
`const`

rniwa replied:
Fixed.

smfr commented on Source/WebCore/dom/EventLoop.h:
```diff
@@ -119,16 +121,19 @@ class EventLoop : public RefCounted<EventLoop>, public CanMakeWeakPtr<EventLoop>
     void addAssociatedContext(ScriptExecutionContext&);
     void removeAssociatedContext(ScriptExecutionContext&);
 
+    void invalidateNextTimerFiringTimeCache() { m_nextTimerFiringTimeCache = std::nullopt; }
```
"firing time" -> "fire time" for these two names.

rniwa replied:
Fixed.

smfr commented on Source/WebCore/dom/IdleCallbackController.h:
```diff
@@ -46,6 +46,7 @@ class IdleCallbackController : public CanMakeWeakPtr<IdleCallbackController> {
     void removeIdleCallback(int);
 
     void startIdlePeriod();
+    bool isEmpty() { return m_idleRequestCallbacks.isEmpty() && m_runnableIdleCallbacks.isEmpty(); }
```
`const`

rniwa replied:
Fixed.

whsieh commented on Source/WebCore/dom/IdleCallbackController.h:
```diff
@@ -46,6 +46,7 @@ class IdleCallbackController : public CanMakeWeakPtr<IdleCallbackController> {
     void removeIdleCallback(int);
 
     void startIdlePeriod();
+    bool isEmpty() { return m_idleRequestCallbacks.isEmpty() && m_runnableIdleCallbacks.isEmpty(); }
```
Nit - `const` here too.

rniwa replied:
Fixed.

whsieh commented on Source/WebCore/dom/EventLoop.cpp:
```diff
@@ -39,6 +39,7 @@ class EventLoopTimer final : public RefCounted<EventLoopTimer>, public TimerBase
 
     Type type() const { return m_type; }
     EventLoopTaskGroup* group() const { return m_task ? m_task->group() : nullptr; }
+    bool isSuspended() { return m_suspended; }
```
Nit - `const`.

rniwa replied:
Fixed.

smfr commented on Source/WebCore/dom/WindowEventLoop.cpp:
```diff
@@ -158,21 +158,52 @@ bool WindowEventLoop::shouldEndIdlePeriod()
+    auto now = MonotonicTime::now();
+
+    auto timerTime = nextTimerFiringTime();
+    if (timerTime && *timerTime <= now + 4_ms) {
```
Could you move the 4_ms into a named variable to explain how its was derived?

rniwa replied:
I made it up out of thin air. Added IdleCallbackDurationExpectation.

smfr commented on Source/WebCore/dom/WindowEventLoop.cpp:
```diff
@@ -158,21 +158,52 @@ bool WindowEventLoop::shouldEndIdlePeriod()
 
 MonotonicTime WindowEventLoop::computeIdleDeadline()
 {
+    auto minTime = m_lastIdlePeriodStartTime + 50_ms;
```
Where does the 50_ms come from?

rniwa replied:
It's in the spec. Unfortunately spec doesn't have a proper URL for this section of code.

smfr commented on Source/WebCore/page/Page.cpp:
```diff
@@ -1690,7 +1690,11 @@ void Page::scheduleRenderingUpdateInternal()
     forEachWindowEventLoop([&](WindowEventLoop& windowEventLoop) {
-        windowEventLoop.didScheduleRenderingUpdate(*this, m_lastRenderingUpdateTimestamp + preferredRenderingUpdateInterval());
+        auto nextRenderingUpdateTime = m_lastRenderingUpdateTimestamp + preferredRenderingUpdateInterval();
+        auto now = MonotonicTime::now();
```
Can we move this outside the loop?

rniwa replied:
Sure, fixed.

whsieh commented on Source/WebCore/dom/WindowEventLoop.cpp:
```diff
@@ -158,21 +158,52 @@ bool WindowEventLoop::shouldEndIdlePeriod()
+        if (!idleCallbackController)
+            return;
+        hasIdleCallbacks = hasIdleCallbacks || !idleCallbackController->isEmpty();
+    });
```
Nit - I suppose something like a `findAssociatedContextMatching(…)` might be _ever so slightly_ more efficient here, since we can stop iteration early once we find non-empty idle callbacks (though the list of associated contexts is likely pretty short anyways).

rniwa replied:
Sure, added findMatchingAssociatedContext to EventLoop.cpp.

whsieh commented on Source/WebCore/dom/IdleDeadline.cpp:
```diff
@@ -38,9 +38,9 @@ DOMHighResTimeStamp IdleDeadline::timeRemaining(Document& document) const
-    auto deadline = document.windowEventLoop().computeIdleDeadline();
-    auto remainingTime = window->performance().relativeTimeFromTimeOriginInReducedResolution(deadline) - window->performance().now();
-    return remainingTime < 0 ? 0 : remainingTime;
+    auto duration = document.windowEventLoop().computeIdleDeadline() - MonotonicTime::now();
```
While this `MonotonicTime::now()` is not new since it came from `Performance::nowInReducedResolutionSeconds()`, I wonder if we could use `ApproximateTime` instead?

rniwa replied:
We can't use ApproximateTime for this since this value will be exposed to JS via IdleDeadline.timeRemaining.

rniwa committed:
Event loop should attempt to run idle callbacks at the end of each task
https://bugs.webkit.org/show_bug.cgi?id=260115

Reviewed by Wenson Hsieh.

This PR implements the step 8 in the HTML5 event loop processing model:
https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model

Specifically, after executing each task in WindowEventLoop, we check if there are any tasks or microtasks
scheduled in the event loop. If there are none, we attempt to run idle callbacks.

This PR also fixes IdleDeadline::timeRemaining to subtract the current time before reducing the resolution
as it turns out that reducing the resolution before subtraction can generate the remaining time that is
slightly longer than the actual time remaining. e.g. 50.000000002ms as opposed to 50ms.

In addition, this PR updates WindowEventLoop::computeIdleDeadline() to take into account the next time
a timer associated with this event loop fires. This is computed & cached by EventLoop::nextTimerFiringTime.

Finally, this PR also fixes a bug in Page::scheduleRenderingUpdateInternal() that it was sometimes calling
WindowEventLoop::didScheduleRenderingUpdate with a rendering update time in the past.

* LayoutTests/http/tests/eventloop/idle-callback-not-affected-by-timers-in-cross-origin-expected.txt: Added.
* LayoutTests/http/tests/eventloop/idle-callback-not-affected-by-timers-in-cross-origin.html: Added.
* LayoutTests/http/tests/eventloop/resources/idle-callback-timer-helper.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/requestidlecallback/deadline-max-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/requestidlecallback/deadline-max-rAF-dynamic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/requestidlecallback/deadline-max-timeout-dynamic-expected.txt:
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/requestidlecallback/requestidlecallback-in-page-cache.html: This test was relying on the idle
callback in the main frame to be invoked after idle callbacks in subframes. Since this is not guaranteed,
* Source/WebCore/dom/EventLoop.cpp:
(WebCore::EventLoop::scheduleTask):
(WebCore::EventLoop::removeScheduledTimer):
(WebCore::EventLoop::scheduleRepeatingTask):
(WebCore::EventLoop::removeRepeatingTimer):
(WebCore::EventLoop::findMatchingAssociatedContext): Added.
(WebCore::EventLoop::nextTimerFiringTime const): Added.
(WebCore::EventLoopTaskGroup::suspend):
(WebCore::EventLoopTaskGroup::resume):
(WebCore::EventLoopTaskGroup::setTimerAlignment):
(WebCore::EventLoopTaskGroup::didChangeTimerAlignmentInterval):
(WebCore::EventLoopTaskGroup::setTimerHasReachedMaxNestingLevel):
(WebCore::EventLoopTaskGroup::adjustTimerNextFireTime):
(WebCore::EventLoopTaskGroup::adjustTimerRepeatInterval):
* Source/WebCore/dom/EventLoop.h:
(WebCore::EventLoop::invalidateNextTimerFiringTimeCache): Added.
* Source/WebCore/dom/IdleCallbackController.h:
(WebCore::IdleCallbackController::isEmpty):
* Source/WebCore/dom/IdleDeadline.cpp:
(WebCore::IdleDeadline::timeRemaining const):
* Source/WebCore/dom/WindowEventLoop.cpp:
(WebCore::WindowEventLoop::computeIdleDeadline):
(WebCore::WindowEventLoop::didReachTimeToRun):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::scheduleRenderingUpdateInternal):
* Source/WebCore/platform/Timer.h:
(WebCore::TimerBase::nextFireTime const):

Canonical link: https://commits.webkit.org/266877@main
